File: Assets/Scripts/Entity/XEntityMgr.cs
using UnityEngine;
using System.Collections.Generic;
using System;
using XTable;

public class XEntityMgr : XSingleton<XEntityMgr>
{
    private List<XEntity> _empty = new List<XEntity>();
    private Dictionary<uint, XEntity> _dic_entities = new Dictionary<uint, XEntity>();
    private HashSet<XEntity> _hash_entitys = new HashSet<XEntity>();
    private Dictionary<EntityType, List<XEntity>> _map_entities = new Dictionary<EntityType, List<XEntity>>();
    public XPlayer Player;

    private T PrepareEntity<T>(XAttributes attr) where T : XEntity
    {
        T x = Activator.CreateInstance<T>();
        GameObject o = XResources.LoadInPool("Prefabs/" + attr.Prefab);
        o.name = attr.id.ToString();
        o.transform.position = attr.AppearPostion;
        o.transform.rotation = attr.AppearQuaternion;
        x.Initilize(o, attr);
        if (!_dic_entities.ContainsKey(attr.id) && !IsPlayer(x)) _dic_entities.Add(attr.id, x);
        if (!_hash_entitys.Add(x) && !IsPlayer(Player)) XDebug.Log("has exist entity: ", attr.id);
        SetRelation(attr.id, attr.FightGroup);
        return x;
    }


    public void UnloadAll()
    {
        var e = _hash_entitys.GetEnumerator();
        while (e.MoveNext())
        {
            e.Current.UnloadEntity();
        }
        _hash_entitys.Clear();
        _dic_entities.Clear();
        _map_entities.Clear();
    }


    public void UnloadEntity<T>(uint id)
    {
        if (_dic_entities.ContainsKey(id))
        {
            _dic_entities[id].UnloadEntity();
            _dic_entities.Remove(id);
        }
        _hash_entitys.RemoveWhere(x => x.EntityID == id);
        var e = _map_entities.GetEnumerator();
        while (e.MoveNext())
        {
            e.Current.Value.RemoveAll(ent => ent.EntityID == id);
        }
    }

    public void Update(float delta)
    {
        var e = _hash_entitys.GetEnumerator();
        while (e.MoveNext())
        {
            e.Current.OnUpdate(delta);
        }
    }


    public void LateUpdate()
    {
        var e = _hash_entitys.GetEnumerator();
        while (e.MoveNext())
        {
            e.Current.OnLateUpdate();
        }
    }


    /// <summary>
    /// 登录进入
    /// </summary>
    public void AttachToHost()
    {
        XAttachEventArgs e = new XAttachEventArgs();
        XEventMgr.singleton.FireEvent(e);
    }

    /// <summary>
    /// 退出登录
    /// </summary>
    public void DetachFromHost()
    {
        XDetachEventArgs e = new XDetachEventArgs();
        XEventMgr.singleton.FireEvent(e);
    }

    public XEntity GetEntity(uint id)
    {
        if (_dic_entities.ContainsKey(id))
            return _dic_entities[id];
        return null;
    }


    public XEntity CreateEntity<T>(uint staticid, Vector3 pos, Quaternion rot) where T : XEntity
    {
        XAttributes attr = InitAttrFromClient((int)staticid);
        attr.AppearPostion = pos;
        attr.AppearQuaternion = rot;
        XEntity e = PrepareEntity<T>(attr);
        Add(EntityType.Entity, e);
        return e;
    }

    public XRole CreateTestRole()
    {
        XAttributes attr = InitAttrFromClient(2);
        attr.AppearPostion = Vector3.zero;
        attr.AppearQuaternion = Quaternion.identity;
        return PrepareEntity<XRole>(attr);
    }

    public XPlayer CreatePlayer()
    {
        SceneList.RowData row = XScene.singleton.SceneRow;
        int statisticid = 2;
        XAttributes attr = InitAttrFromClient(statisticid);
        string s = row.StartPos;
        string[] ss = s.Split('=');
        float[] fp = new float[3];
        float.TryParse(ss[0], out fp[0]);
        float.TryParse(ss[1], out fp[1]);
        float.TryParse(ss[2], out fp[2]);
        attr.AppearPostion = new Vector3(fp[0], fp[1], fp[2]);
        attr.AppearQuaternion = Quaternion.Euler(row.StartRot[0], row.StartRot[1], row.StartRot[2]);
        return Player = PrepareEntity<XPlayer>(attr);
    }

    public XNPC CreateNPC(XNpcList.RowData row)
    {
        var pos = new Vector3(row.NPCPosition[0], row.NPCPosition[1], row.NPCPosition[2]);
        var rot = Quaternion.Euler(row.NPCRotation[0], row.NPCRotation[1], row.NPCRotation[2]);
        return CreateNPC(row, pos, rot);
    }

    public XNPC CreateNPC(XNpcList.RowData row, Vector3 pos, Quaternion rot)
    {
        XAttributes attr = InitAttrByPresent(row.PresentID);
        attr.AppearPostion = pos;
        attr.AppearQuaternion = rot;
        var e = PrepareEntity<XNPC>(attr);
        Add(EntityType.Npc, e);
        return e;
    }

    public List<XEntity> GetAllNPC()
    {
        return _map_entities[EntityType.Npc];
    }

    public List<XEntity> GetAllAlly()
    {
        return _map_entities[EntityType.Ally];
    }

    public List<XEntity> GetAllAlly(XEntity e)
    {
        EntityType type = (EntityType)e.Attributes.FightGroup;
        if (type == EntityType.Ally) return _map_entities[EntityType.Ally];
        else if (type == EntityType.Enemy) return _map_entities[EntityType.Enemy];
        return _empty;
    }

    public List<XEntity> GetAllEnemy()
    {
        return _map_entities[EntityType.Enemy];
    }

    public List<XEntity> GetAllEnemy(XEntity e)
    {
        EntityType type = (EntityType)e.Attributes.FightGroup;
        if (type == EntityType.Ally) return _map_entities[EntityType.Enemy];
        else if (type == EntityType.Enemy) return _map_entities[EntityType.Ally];
        return _empty;
    }

    public void SetRelation(uint entityid, int group)
    {
        int max = EntityType.Ship_End - EntityType.Ship_Start;
        if (group < 0 || group > max)
        {
            XDebug.LogError("Set Relation err ", entityid);
            return;
        }
        EntityType type = group + EntityType.Ship_Start;
        XEntity e = null;
        if (_dic_entities.TryGetValue(entityid, out e))
        {
            e.SetRelation(type);
            Add(type, e);
        }
    }

    private bool IsPlayer(XEntity e)
    {
        if (Player != null)
            return Player.EntityID == e.EntityID;
        return false;
    }

    private XAttributes InitAttrFromClient(int statisticid)
    {
        var srow = XTableMgr.GetTable<XEntityStatistics>().GetByID(statisticid);
        if (srow == null) throw new Exception("entity is nil with id: " + statisticid);
        XAttributes attr = new XAttributes();
        var prow = XTableMgr.GetTable<XEntityPresentation>().GetItemID(srow.PresentID);
        if (prow == null) throw new Exception("present is nil with id: " + srow.PresentID);
        attr.Prefab = prow.Prefab;
        attr.id = (uint)XCommon.singleton.New_id;
        attr.PresentID = srow.PresentID;
        attr.Name = prow.Name;
        attr.InitAttribute(srow);
        return attr;
    }

    private XAttributes InitAttrByPresent(uint presentID)
    {
        XAttributes attr = new XAttributes();
        var prow = XTableMgr.GetTable<XEntityPresentation>().GetItemID(presentID);
        if (prow == null) throw new Exception("present is nil with id: " + presentID);
        attr.Prefab = prow.Prefab;
        attr.id = (uint)XCommon.singleton.New_id;
        attr.PresentID = presentID;
        attr.Name = prow.Name;
        return attr;
    }

    private bool Add(EntityType type, XEntity e)
    {
        if (_map_entities.ContainsKey(type))
        {
            List<XEntity> l = _map_entities[type];
            if (l.Contains(e)) return false;
            else l.Add(e);
        }
        else
        {
            List<XEntity> l = new List<XEntity>();
            l.Add(e);
            _map_entities.Add(type, l);
        }
        return true;
    }

}

File: tools_proj/XCPP/GameCore/XEntityMgr.cpp
#include "XEntityMgr.h"
#include "XEntity.h"
#include "XAttributes.h"
#include "XPlayer.h"
#include "XNpc.h"
#include "XEntityStatistics.h"
#include "XEntityPresentation.h"
#include "XScene.h"

XEntityMgr::~XEntityMgr()
{
}

void XEntityMgr::Update(float delta)
{
	for (std::unordered_set<XEntity*>::iterator itr = _hash_entities.begin(); itr != _hash_entities.end(); itr++)
	{
		(*itr)->Update(delta);
	}
	if (Player)
	{
		Player->Update(delta);
	}
}

void XEntityMgr::AttachToHost()
{
	for (std::unordered_set<XEntity*>::iterator itr = _hash_entities.begin(); itr != _hash_entities.end(); itr++)
	{
		(*itr)->AttachToHost();
	}
}

void XEntityMgr::DetachFromHost()
{
	for (std::unordered_set<XEntity*>::iterator itr = _hash_entities.begin(); itr != _hash_entities.end(); itr++)
	{
		(*itr)->DetachFromHost();
	}
}

 void XEntityMgr::InitAttr(int staticid, XAttributes* attr)
{
	XEntityStatisticsRow srow;
	iGetXEntityStatisticsRowByID(staticid, &srow);
	XEntityPresentationRow prow;
	iGetXEntityPresentationRowByID((uint)srow.presentid, &prow);
	attr->setid(new_id());
	attr->setPresentID(srow.presentid);
	attr->setPrefab(prow.prefab);
	attr->setName(prow.name);
	attr->setAIBehaviour(srow.aibehavior);
}


 std::vector<XEntity*> XEntityMgr::GetAllEnemy(XEntity* e)
 {
	 EntityType type = (EntityType)(1 << (e->getAttributes()->FightGroup + Ship_Start));
	 if (type == Ally) return _map_entities[Enemy];
	 else if (type == Enemy) return _map_entities[Ally];
	 return _empty;
 }


 std::vector<XEntity*> XEntityMgr::GetAllAlly(XEntity* e)
 {
	 EntityType type = (EntityType)(1 << (e->getAttributes()->FightGroup + Ship_Start));
	 if (type == Ally) return _map_entities[Enemy];
	 else if (type == Enemy) return _map_entities[Ally];
	 return _empty;
 }

 std::vector<XEntity*> XEntityMgr::GetAllNPC()
 {
	 return _map_entities[Npc];
 }

void XEntityMgr::PrepareEntity(XAttributes* attr, XEntity* ent)
{
	ent->setAttributes(attr);
	std::string str = "Prefabs/" + tostring(attr->getPrefab());
	GameObject* o = GameObjectMgr::Instance()->Create(str.c_str());
	o->name = tostring(attr->getid()).c_str();
	o->transform->position = attr->getAppearPostion();
	Vector3 rot = attr->getAppearQuaternion();
	o->transform->rotation = rot;

	ent->Initilize(o, attr);
	ent->EntityID = attr->getid();
	if (!ent->IsPlayer())
		_dic_entities.insert(std::make_pair(id, ent));
}

bool XEntityMgr::Add(EntityType type, XEntity* e)
{
	std::vector<XEntity*> vec = _map_entities[(int)type];
	bool exit = false;
	for (size_t i = 0; i < vec.size(); i++)
	{
		if (vec[i] == e) exit = true;
	}
	if (exit) vec.push_back(e);
	_hash_entities.insert(e);
	return !exit;
}

void XEntityMgr::UnloadEntity(uint id)
{
	std::unordered_map<uint, XEntity*>::iterator itr = _dic_entities.find(id);
	if (itr != _dic_entities.end())
	{
		_dic_entities.erase(id);
		_hash_entities.erase(itr->second);
		delete itr->second->getAttributes();
		delete itr->second;
	}
	std::unordered_map<int, std::vector<XEntity*>>::iterator it;
	for (it = _map_entities.begin(); it != _map_entities.end(); it++)
	{
		std::vector<XEntity*> vec = it->second;
		std::vector<XEntity*>::iterator vtr;
		for (vtr = vec.begin(); vtr != vec.end();)
		{
			if ((*vtr)->EntityID == id)
			{
				vec.erase(vtr);
			}
			else
			{
				vtr++;
			}
		}
	}
}

void XEntityMgr::UnloadAll()
{
	std::unordered_set<XEntity*>::iterator itr;
	for (itr = _hash_entities.begin(); itr != _hash_entities.end(); itr++)
	{
		delete (*itr)->getAttributes();
		delete *itr;
	}
	_hash_entities.clear();
	_map_entities.clear();
	_dic_entities.clear();

	if (Player)
	{
		delete Player->getAttributes();
		delete Player;
	} 
	Player = NULL;
}


XEntity* XEntityMgr::GetEntity(uint id)
{
	if (Player && id == Player->EntityID)
	{
		return Player;
	}
	else
	{
		return _dic_entities[id];
	}
}


XPlayer* XEntityMgr::CreatePlayer()
{
	int staticid = 2;
	XAttributes* attr = new XAttributes();
	InitAttr(staticid, attr);
	SceneListRow* row = XScene::Instance()->getSceneRow();
	if (row)
	{
		std::vector<std::string> s = split(row->startpos, '=');
		float x = (float)atof(s[0].c_str());
		float y = (float)atof(s[1].c_str());
		float z = (float)atof(s[2].c_str());
		attr->setAppearPosition(Vector3(x, y, z));
		attr->setAppearQuaternion(Vector3(row->startrot[0], row->startrot[1], row->startrot[2]));
	}
	Player = new XPlayer();
	PrepareEntity(attr, Player);
	return Player;
}


XEntity* XEntityMgr::CreateNPC(XNpcListRow& row, Vector3 pos, Vector3 rot)
{
	XAttributes* attr = InitAttrByPresent(row.presentid);
	attr->setAppearPosition(pos);
	attr->setAppearQuaternion(rot);
	XEntity* enty = new XNpc();
	PrepareEntity(attr, enty);
	Add(Npc, enty);
	return enty;
}

XAttributes* XEntityMgr::InitAttrByPresent(uint presentID)
{
	XAttributes* attr = new XAttributes();
	XEntityPresentationRow row;
	iGetXEntityPresentationRowByID(presentID, &row);
	attr->setPresentID(presentID);
	attr->setPrefab(row.prefab);
	attr->setid(new_id());
	attr->setName(row.name);
	return attr;
}